python/tests/utils.py

```diff
@@ -13,3 +13,16 @@ def compare_arrays(test_instance, exp, res, tol=1e-3):
     else:
         diff = np.linalg.norm(res_1d - exp_1d) / norm_exp
         test_instance.assertLess(diff, tol)
+
+
+class VectorComparisonMixin(unittest.TestCase):
+    """A mixin for adding proper floating point value and vector comparison."""
+
+    def assertVectorsClose(self, x, y, epsilon = 1e-2, msg = ''):
```

Comment:
It would be useful to be consistent with the function naming convention: all--lower-case words separated by an underscore; i.e. `assert_vec_close` as suggested in #1575. (This is consistent with the other function defined this file `compare_arrays`.)

Also, why does this function need to be part of a mixin class when its arguments are only of a single type: arrays of floating-point numbers?

Reply:
The goal of going the mixin route was to use the `assertLessEqual()` method of `unittest.TestCase`, which tends to have better reporting and testing framework integration than using a vanilla `assert`. This strikes me as being cleaner than the approach used by `compare_arrays()` which takes `test_instance` but isn't a proper method of the test case class. The camel case naming in `assertVectorsClose()` matches the convention of the other methods in the `TestCase` class.

Reply:
Looking at `compare_arrays()`, it actually seems to be almost the same comparison that I've implemented... I didn't actually look at it in detail... I just saw this file and dumped my class in here so it might be reused by the other gradient tests in the future. I actually wonder why the other gradient tests don't already use `compare_arrays()`?

Reply:
Yes, rather than add a new function/class for comparing two arrays as in this PR, it might make more sense to simply revamp `compare_arrays` to switch from an L<sub>2</sub> to L<sub>∞</sub> norm and update the affected tests in a separate PR.